### PR TITLE
Sandeep's Assignment Submission

### PR DESCRIPTION
**Pull request by sandeeppal1991**

### CODE REVIEW

File: HW03_ex06.py [hunk 1]
```diff
@@ -4,11 +4,15 @@
...
 # (3) Please remove development code before submitting.
 ###############################################################################
+import math
 # Exercise 6.2
 # See 6.1: "write a compare function takes two values, x and y, and returns 1
 # if x > y, 0 if x == y, and -1 if x < y."
 # When you submit only include your final function: compare
-
+def compare(x,y):
+    if(x>y):return 1
```
Comment (danielsgriffin) [inline HW03_ex06.py]: You will see in line conditional execution so *very* rarely in python. 

Also note that you don't need the parenthesis.

File: HW03_ex06.py [hunk 2]
```diff
@@ -22,7 +26,8 @@
 # Do develop incrementally. Do not share in your final push your incremental
 # work.
 
-
+def hypotenuse(a,b):
+    return math.sqrt(a**2+b**2)
```
Comment (danielsgriffin) [inline HW03_ex06.py]: fyi: `math.sqrt(a**2 + b**2)` == `(a**2 + b**2)**.5` (w/ no import req.)

File: HW03_ex06.py [hunk 3]
```diff
@@ -22,7 +26,8 @@
 # Do develop incrementally. Do not share in your final push your incremental
 # work.
 
-
+def hypotenuse(a,b):
```
Comment (danielsgriffin) [inline HW03_ex06.py]: `f(p1,p2) => f(p1, p2)`

File: HW03_ex06.py [hunk 4]
```diff
@@ -32,7 +37,10 @@
 # or False otherwise"
 # When you submit only include your final function: is_between
 
-
+def is_between(x,y,z):
```
Comment (danielsgriffin) [inline HW03_ex06.py]: `f(p1,p2) => f(p1, p2)` add spaces

File: HW03_ex06.py [hunk 5]
```diff
@@ -32,7 +37,10 @@
...
+    if(y>=x and y <= z):
```
Comment (danielsgriffin) [inline HW03_ex06.py]: add spaces around conditional operators: `y >= x `

Comment (danielsgriffin) [inline HW03_ex06.py]: note you can chain the comparisons:
 `x <= y <= z` == `(x <= y) and (y <= z)`
    https://docs.python.org/3/reference/expressions.html#comparisons

File: HW03_ex06.py [hunk 6]
```diff
@@ -58,6 +72,12 @@
 # (whole numbers not including zero)
 # When you submit only include your final function: is_power
 
+def is_power(a,b):
+
+    if(a!= 1.0):
```
Comment (danielsgriffin) [inline HW03_ex06.py]: have the same space on both sides of the != for readability/convention

File: HW03_ex06.py [hunk 7]
```diff
@@ -58,6 +72,12 @@
...
+        return (a%b == 0 and is_power(a/b,b))
```
Comment (danielsgriffin) [inline HW03_ex06.py]: you don't need the parenthesis unless for readability

Comment (danielsgriffin) [inline HW03_ex06.py]: `f(p1,p2) => f(p1, p2)`

File: HW03_ex06.py [hunk 8]
```diff
@@ -58,6 +72,12 @@
...
+    else:
+        return True
```
Comment (danielsgriffin) [inline HW03_ex06.py]: Is 1 a power of 2? Will your code say so?

File: HW03_string1.py [hunk 9]
```diff
@@ -25,8 +25,13 @@
 # So donuts(5) returns 'Number of donuts: 5'
 # and donuts(23) returns 'Number of donuts: many'
 def donuts(count):
-    # +++your code here+++
-    return
+    #simple if else statement... value is many if more than or equal to 10
+    #else the statement is printed
+    if(count >= 10):
```
Comment (danielsgriffin) [inline HW03_string1.py]: parens not necessary

File: HW03_string1.py [hunk 10]
```diff
@@ -25,8 +25,13 @@
...
+        donuts_value = "many"
+    else:
+        donuts_value = str(count)
+    return "Number of donuts: " + donuts_value
```
Comment (danielsgriffin) [inline HW03_string1.py]: good use of variables

File: HW03_string1.py [hunk 11]
```diff
@@ -35,8 +40,13 @@ def donuts(count):
 # so 'spring' yields 'spng'. However, if the string length
 # is less than 2, return instead the empty string.
 def both_ends(s):
-    # +++your code here+++
-    return
+    #I have compared the string length and if less than 2 then the variable gets
+    #assigned no value else i have used slicing to modify the string
+    if(len(s) < 2):
+        both_ends_value = ""
```
Comment (danielsgriffin) [inline HW03_string1.py]: calling this the variable name feels deceptive semantically to the reader though it functionally runs. you are NOT returning both ends if it is < 2.

File: HW03_string1.py [hunk 12]
```diff
@@ -49,8 +59,9 @@ def both_ends(s):
 # Hint: s.replace(stra, strb) returns a version of string s
 # where all instances of stra have been replaced by strb.
 def fix_start(s):
-    # +++your code here+++
-    return
+    #here i have replaced the first letter by * in the sliced version of the
+    #string. I then added it to the first character
+    return s[0] + s[1:].replace(s[0],'*')
```
Comment (danielsgriffin) [inline HW03_string1.py]: beautiful

File: HW03_string1.py [hunk 13]
```diff
@@ -61,8 +72,9 @@ def fix_start(s):
 #   'dog', 'dinner' -> 'dig donner'
 # Assume a and b are length 2 or more.
 def mix_up(a, b):
-    # +++your code here+++
-    return
+    # replacing substring of a with b and replacing substring of b with a
+    # and then adding them together
+    return a.replace(a[:2],b[:2]) +" "+ b.replace(b[:2],a[:2])
```
Comment (danielsgriffin) [inline HW03_string1.py]: Interesting, if complex. But though it works for the test cases given, it is not correct.

Ex.:
`print(mix_up("haha", "sorry"))`

**Comment by danielsgriffin** [comment]
Good job.

File: HW03_ex06.py [hunk 14]
```diff
@@ -45,6 +53,12 @@
 # When you submit only include your final function: is_palindrome
 
 
+def is_palindrome(str):
+    str = str.lower()
+    if(str[::-1] == str):
```
Comment (sandeeppal1991) [inline HW03_ex06.py]: My Bad. I thought a solution with lesses space time complexity would be ideal. Will keep in mind instructions from next time. Thanks.

Comment (danielsgriffin) [inline HW03_ex06.py]: No worries. I didn't originally give instructions to do it recursively.

**Pull request closed without merging**